src/dragonball/docs/upcall.md

```diff
@@ -0,0 +1,30 @@
+# `Upcall`
+
+## What is `Upcall`?
+
+`Upcall` is a direct communication tool between VMM and guest developed upon `vsock`. The server side of the `upcall` is a driver in guest kernel (kernel patches are needed for this feature) and it'll start to serve the requests after the kernel starts. And the client side is in Dragonball VMM , it'll be a thread that communicates with `vsock` through `uds`.
+
+We want to keep the lightweight of the VM through the implementation of the `upcall`.
+
+![architecture overview](images/upcall-architecture.svg)
+## What can `upcall` do?
+
+We define specific operations in the device manager service (one of the services in `upcall` we developed) to perform device hotplug / hot-unplug including vCPU hotplug, `virtio-mmio` hotplug, and memory hotplug. We have accomplished device hotplug / hot-unplug directly through `upcall` in order to avoid the virtualization of ACPI to minimize virtual machines overhead. And there could be many other uses if other services are implemented.
+
+## How to enable `upcall`?
+
+`Upcall` needs a server in the guest kernel which will be several kernel patches for the `upcall` server itself and different services registered in the `upcall` server. It's currently tested on upstream Linux kernel 5.10.
+
+To make it easy for users to use, we have open-source the `upcall` guest patches in [Dragonball experimental guest patches](../../../tools/packaging/kernel/patches/5.10.x/dragonball-experimental) and develop `upcall` support in [Kata guest kernel building script](../../../tools/packaging/kernel/build-kernel.sh).
+
+You could use following command to download the upstream kernel (currently Dragonball uses 5.10.25) and put the `upcall` patches and other Kata patches into kernel code.
+
+`sh build-kernel.sh -e -t dragonball -f setup`
```

Comment:
Sorry for being late in the party, but I think we should have a single experimental kernel. It means that if we ever need other experimental kernel patches, we can put them all together instead of having to set up an experimental kernel for each different features. wdyt? /cc @fidencio @egernst 

btw, for upcall, it might actually be useful for other hypervisors too, e.g., firecracker?

Reply:
Hi @bergwolf, sorry that I just notice your comments.
1. Right now, we will put all Dragonball experimental features (right now only upcall) into Dragonball experimental guest kernel. I am not quite familiar with other experimental kernel features but I think those patches may target different kernel version so it will be kinda hard to put all experimental features into just one experimental kernel 

2. Yes, ideally firecracker could use upcall to do things like cpu hot plug but they need to implement upcall client side in the VMM first.

Reply:
@studychao, fair enough. I took a look at the `tools/packaging/kernel` directory and it has evolved a lot since the last time I looked. IMO it is quite a burden to maintain so many kernel patches and versions separately but looks like folks are just fine with it. So let's live with it ;)